client/app/components/SearchBar/SearchBar.js

```diff
@@ -0,0 +1,76 @@
+import { Component, PropTypes } from 'react';
+import { form, input, button, span } from 'r-dom';
+
+import css from './SearchBar.css';
+import icon from './images/search-icon.svg';
+
+const SEARCH_MODE_KEYWORD = 'keyword';
+const SEARCH_MODE_LOCATION = 'location';
+const SEARCH_MODE_KEYWORD_AND_LOCATION = 'keyword-and-location';
+const SEARCH_MODES = [
+  SEARCH_MODE_KEYWORD,
+  SEARCH_MODE_LOCATION,
+  SEARCH_MODE_KEYWORD_AND_LOCATION,
+];
+
+class SearchBar extends Component {
+  constructor(props) {
+    super(props);
+
+    // The values of the inputs are synced to the component state
+    // until the form is submitted.
+    this.state = {
+      keyword: '',
+      location: '',
+    };
+  }
+  handleSubmit(e) {
+    e.preventDefault();
+    this.props.onSubmit(this.state);
+  }
+  render() {
+    const { mode, keywordPlaceholder, locationPlaceholder } = this.props;
+
+    const keywordInput = input({
+      type: 'search',
+      className: css.keywordInput,
+      placeholder: keywordPlaceholder,
+      onChange: (e) => this.setState({ keyword: e.target.value }), // eslint-disable-line react/no-set-state
+    });
+    const locationInput = input({
+      type: 'search',
+      className: css.locationInput,
+      placeholder: locationPlaceholder,
+      autoComplete: 'off',
+      onChange: (e) => this.setState({ location: e.target.value }), // eslint-disable-line react/no-set-state
+    });
+
+    const hasKeywordInput = mode === SEARCH_MODE_KEYWORD || mode === SEARCH_MODE_KEYWORD_AND_LOCATION;
+    const hasLocationInput = mode === SEARCH_MODE_LOCATION || mode === SEARCH_MODE_KEYWORD_AND_LOCATION;
+
+    return form({
+      classSet: {
+        [css.root]: true,
+      },
+      onSubmit: this.handleSubmit.bind(this),
```

Comment:
I've usually kept a habit of writing 'this' bindings in constructor. It's a good place to do bindings since functions can be used in several places - and having bindings all over the code seems a bit noisy/unclear for me.